test/jdk/java/foreign/TestUpcallException.java

```diff
@@ -0,0 +1,83 @@
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @requires ((os.arch == "amd64" | os.arch == "x86_64") & sun.arch.data.model == "64") | os.arch == "aarch64"
+ * @library /test/lib
+ * @modules jdk.incubator.foreign/jdk.internal.foreign
+ * @build ThrowingUpcall TestUpcallException
+ *
+ * @run testng/othervm/native
+ *   --enable-native-access=ALL-UNNAMED
+ *   TestUpcallException
+ */
+
+import jdk.test.lib.Utils;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.concurrent.ThreadLocalRandom;
+
+import static org.testng.Assert.assertEquals;
+
+public class TestUpcallException {
+
+    @Test
+    public void testExceptionInterpreted() throws InterruptedException, IOException {
+        boolean useSpec = false;
+        run(useSpec);
+    }
+
+    @Test
+    public void testExceptionSpecialized() throws IOException, InterruptedException {
+        boolean useSpec = true;
+        run(useSpec);
+    }
+
+    private void run(boolean useSpec) throws IOException, InterruptedException {
+        int exitCode = ThreadLocalRandom.current().nextInt();
+
+        Process process = new ProcessBuilder()
+            .inheritIO()
+            .command(
+                Paths.get(Utils.TEST_JDK)
+                     .resolve("bin")
+                     .resolve("java")
+                     .toAbsolutePath()
+                     .toString(),
+                "--add-modules", "jdk.incubator.foreign",
+                "--enable-native-access=ALL-UNNAMED",
+                "-Djava.library.path=" + System.getProperty("java.library.path"),
+                "-Djdk.internal.foreign.ProgrammableUpcallHandler.USE_SPEC=" + useSpec,
+                "-Djdk.incubator.foreign.uncaught_exception_code=" + exitCode,
+                "-cp", Utils.TEST_CLASS_PATH,
+                "ThrowingUpcall")
+            .start();
+
+        int result = process.waitFor();
+        assertEquals(result, exitCode);
```

Comment:
Should we check for stacktrace output - e.g. the name of ThrowingUpcall" ?

Reply:
That's probably a good idea. Though, I'll note that it still allows for false positives if the tests throws an exception from ThrowingUpcall another way. Comparing the exit codes should catch that though.